cmd/sync-tags/main.go

```diff
@@ -199,13 +214,19 @@ func main() {
 			continue
 		}
 
-		// skip if it already exists in origin
-		if _, found := bTagCommits[bName]; found {
-			continue
-		}
+		// check if the tags exist either at origin or locally
+		_, nonSemverTagAtOrigin := bTagCommits[bName]
+		nonSemverTagExists := nonSemverTagAtOrigin || tagExists(r, bName)
+
+		_, semverTagAtOrigin := bTagCommits[semverTag]
+		semverTagExists := semverTagAtOrigin || tagExists(r, semverTag)
 
-		// do not override tags (we build master first, i.e. the x.y.z-alpha.0 tag on master will not be created for feature branches)
-		if tagExists(r, bName) {
+		// skip tags if they exist locally or at origin already
+		// we want to ensure to not override tags
+		// (we build master first, i.e. the x.y.z-alpha.0 tag on master will not be created for feature branches)
+		if publishSemverTag && semverTagExists && nonSemverTagExists {
+			continue
+		} else if nonSemverTagExists {
 			continue
 		}
```

Comment:
this bit still looks strange to me. if either tag exists at origin, we cannot make changes and should skip, right? something like this, maybe?

```go
if publishSemverTag && semverTagAtOrigin {
  continue
}
if nonSemverTagAtOrigin {
  continue
}
```

for the local checks, I'm not sure how re-entrant this script is expected to be (e.g. if it crashes after creating a local non-semver tag, but before pushing, do we wipe/reclone on next run?)

Reply:
Will let @sttts confirm but this is my understanding:

> if it crashes after creating a local non-semver tag, but before pushing, do we wipe/reclone on next run?

The tag will be [appended in the push script](https://github.com/kubernetes/publishing-bot/blob/23fba6b1430ce9a03fc61bde591c4e1bdbf7eb84/cmd/sync-tags/main.go#L349) once it's created, so it'll try pushing again when the push script is executed on the next run.

>  if either tag exists at origin, we cannot make changes and should skip, right?

Because the script handles cases when tags have been created locally but not pushed, I think we'll need to consider cases when the tag exists locally as well (and not just whether it exists at origin).

> if either tag exists

I thought of skipping tag creation if _one_ of them exists but then considered the following cases...though I'm not entirely sure when we'd hit these cases. Maybe if someone deletes a tag? However, retagging on tag deletion is also a bad idea... :thinking: 

- If `kubernetes-1.x.y` exists but `v0.x.y` does not, the go.mod file for `kubernetes-1.x.y` would point to `v0.x.y` so we **need** to make sure that `v0.x.y` is published.

- If `v0.x.y` exists but `kubernetes-1.x.y` does not, we publish it because both go hand-in-hand (though not publishing it doesn't harm in any way).

Reply:
> The tag will be appended in the push script once it's created, so it'll try pushing again when the push script is executed on the next run.

If I remember right, the push script is recreated on each run (I hope so). Also local tags should be remove before each run.

The mixed logic for semver and non-semver irritates me. I would make all code below idem-potent. Then we could skip this early continue here, but we do it for purely performance reason to skip the git checkout magic if there is no new tag.

Then what @nikhita has in the code looks right. Could be rewritten to the following:

```
if nonSemverTagExists && (!publishSemverTag || semverTagExists) {
  continue
}
```